framework/resource/manager.go

```diff
@@ -0,0 +1,323 @@
+	// function will do nothing, but will take as an input all the marker
+	// values for the resources we want to create. This will force argmapper
+	// to call all our create functions for all our resources.
+	finalInputs := make([]argmapper.Value, 0, len(m.resources))
+	for _, r := range m.resources {
+		finalInputs = append(finalInputs, markerValue(r.name))
+	}
+
+	finalInputSet, err := argmapper.NewValueSet(finalInputs)
+	if err != nil {
+		return err
+	}
+
+	finalFunc, err := argmapper.BuildFunc(
+		finalInputSet, nil,
+		func(in, out *argmapper.ValueSet) error {
+			// no-op on purpose. This function only exists to set the
+			// required inputs for argmapper to create the correct call
+			// graph.
+			return nil
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	// Reset our creation state if we're creating
+	m.createState = &createState{}
+
+	// Start building our arguments
+	mapperArgs, err := m.mapperArgs()
+	if err != nil {
+		return err
+	}
+	for _, arg := range args {
+		mapperArgs = append(mapperArgs, argmapper.Typed(arg))
+	}
+	for _, r := range m.resources {
+		createFunc, err := r.mapperForCreate(m.createState)
+		if err != nil {
+			return err
+		}
+
+		mapperArgs = append(mapperArgs, argmapper.ConverterFunc(createFunc))
+	}
+
+	result := finalFunc.Call(mapperArgs...)
+
+	// If we got an error, perform an automatic rollback.
+	resultErr := result.Err()
+	if resultErr != nil {
+		m.logger.Info("error during creation, starting rollback", "err", resultErr)
+		if err := m.DestroyAll(args...); err != nil {
+			m.logger.Warn("error during rollback", "err", err)
+			resultErr = multierror.Append(resultErr, fmt.Errorf(
+				"Error during rollback: %w", err))
+		} else {
+			m.logger.Info("rollback successful")
+		}
+	}
+
+	return resultErr
+}
+
+// DestroyAll destroys all the resources under management. This will call
+// Destroy in the reverse order of Create. All the state that was created
+// via Create will be available to the Destroy callbacks. Note that after
+// a resource is destroyed, their state is also set to nil.
+//
+// Only resources that have been created will be destroyed. This means
+// that if Create partially failed, then only the resources that attempted
+// creation will have Destroy called. Resources that were never called to
+// Create will do nothing.
+func (m *Manager) DestroyAll(args ...interface{}) error {
+	cs := m.createState
+	if cs == nil || len(cs.Order) == 0 {
+		// We have no creation that was ever done so we have nothing to destroy.
+		return nil
+	}
+
+	var finalInputs []argmapper.Value
+	mapperArgs, err := m.mapperArgs()
+	if err != nil {
+		return err
+	}
+	for _, arg := range args {
+		mapperArgs = append(mapperArgs, argmapper.Typed(arg))
+	}
+
+	// Go through our creation order and create all our destroyers.
+	for i := 0; i < len(cs.Order); i++ {
+		r := m.Resource(cs.Order[i])
+		if r == nil {
+			// We are missing a resource that we should be destroying.
+			return fmt.Errorf(
```

Comment:
Should we destroy the resources that we can, and then return the final error at the end saying what was missed? Or is the assumption that once we hit a missing resource, the rest are also probably in a weird state?

Reply:
Ah nevermind, I see this relates to the Rollback section in the RFC.

Reply:
Yeah honestly this is a weird edge case in this whole thing... I think some experience implementing and iterating on plugins will better inform what our behavior should be here. I could see arguments either way.